src/stdio.rs

```diff
@@ -0,0 +1,93 @@
+//! Enables direct write access to I/O streams in Python's `sys` module.
+
+//! In some cases printing to Rust's `std::io::stdout` or `std::io::stderr` will not appear
+//! in the Python interpreter, e.g. in Jupyter notebooks. This module provides a way to write
+//! directly to Python's I/O streams from Rust in such cases.
+
+//! ```rust
+//! use std::io::Write;
+//!
+//! let mut stdout = pyo3::stdio::stdout();
+//!   
+//! // This may not appear in Jupyter notebooks...
+//! println!("Hello, world!");
+//!
+//! // ...but this will.
+//! writeln!(stdout, "Hello, world!").unwrap();
+//! ```
+
+use crate::ffi::{PySys_WriteStderr, PySys_WriteStdout};
+use crate::Python;
+use std::io::{LineWriter, Write};
+use std::os::raw::c_char;
+
+macro_rules! make_python_stdio {
```

Comment:
I have rewritten it without macros and in favour of a generic implementation of a new `PyStdioRawConfig` trait.   This trait consists entirely of associated constants configuring each stream, with implementations for the structs `PyStdoutRaw` and `PyStderrRaw` which otherwise contain no data.

This is perhaps not ideal, but it was otherwise difficult to get Rust to accept an implementation of the foreign type Write that is generic over local types defined here.   It seems happy with this method because `PyStdioRawConfig`  is "covered".